README.md
# Factorio Resource Objectives

https://mods.factorio.com/mod/ResourceObjectives

Adds intermediate objectives to the game to help get towards rocket launch just a bit easier. Each objective consists of a list of resources, and the player earns additional resources when completing the objectives. Designed for vanilla Factorio, but works with mods as well – especially mods that don't add a lot of new items.

This mod is in development. The objectives have been play-tested only until early oil-era.

Works in multiplayer.

`/next_objective` and `/previous_objective` are handy when upgrading the mod version in the middle of a game.

## Helping with play testing

You can export a log file with `/ro_save_analytics` – this will go to the `script-output` in your factorio folder. The log file contains the in-game timestamps for each completed objective – this is extremely helpful information for balancing the objectives. You can submit an export to [the discussion page](https://mods.factorio.com/mod/ResourceObjectives/discussion/63cdc33313739ac587f9ca39), or to the google forms link below. Log files are useful even if you don't complete the game.

You can also add free-form comments to the log file with `/ro_feedback`. This will add the current in-game timestamp to your message. Especially comments like ”the next objective feels exciting / too easy / too hard” are valuable. You may also explain any milestones that you are reaching in the game outside the explicit objective, or give any sort of ideas.

https://forms.gle/9Ej3QA5seLSK3MiQ6

changelog.txt
---------------------------------------------------------------------------------------------------
Version: 0.6.6
Date: 2023-01-28

  Distribution:
    - Added changelog.txt

---------------------------------------------------------------------------------------------------
Version: 0.6.5
Date: 2023-01-26

  Balancing:
    - Objective #9: Removed 24 stone furnaces (too easy, distracting)
    - Objective #9: Increased transport belts from 100 to 200 (too easy, very needed)
    - Objective #13: Reduced inserters from 500 to 100 (boring, not needed for the factory)
    - Objective #13: Reduced long-handed inserters from 100 to 50 (not needed)
    - Before objective #15: Removed the objective of 100 steel plates. (too easy and following objectives need steel plates anyway)
    - Objective #15: Merged two objectives with 50 assembling machine 2 and 48 steel furnaces. (too easy separately)
    - Objective #24: Moved plastic bar and sulfur before rails
    - Objective #25: Merged rails and rail signal objectives
    - Objective #30: Increased flying robot frames from 100 to 200 (very needed)
    - Objective #31: Increased roboports from 1 to 4 (very needed)

---------------------------------------------------------------------------------------------------
Version: 0.6.4
Date: 2023-01-23

  Balancing:
    - Objective #27: Increased rail chain signals from 10 to 15 (more needed than rail signals)
    - Before objective #30: Removed objective 20 stack inserters (too easy)
    - Before objective #30: Removed objective 50 chemical science packs (boring)
    - Objective #30: Added objective 50 laser turrets and 50 accumulators (military needed, requires a nice amount of research)
    - Before objective #32: Removed objective 100 flamethrower ammo (too easy/late/boring)

---------------------------------------------------------------------------------------------------
Version: 0.6.3
Date: 2023-01-23

  Balancing:
    - Objective #4: Increased gears from 30 to 50 (too easy, very needed)
    - Objective #4: Increased green circuits from 20 to 30 (too easy, very needed)
    - Objective #5: Increased stone bricks from 75 to 100 (too easy)
    - Objective #9: Removed 15 electric miners, 50 small power poles, 5 splitters and 10 underground belts (too hard)
    - Objective #29: Increased red circuits from 500 to 2000 (too easy and needed, 2k is one rail vagon of plastic)

---------------------------------------------------------------------------------------------------
Version: 0.6.2
Date: 2023-01-22

  Features:
    - Added command `/ro_feedback` for saving player comments to analytics log file

  Balancing:
    - Objective #3: Increased coal from 200 to 250 (too easy)
    - Objective #5: Increased stone bricks from 50 to 75 (too easy)
    - Before objective #6: Removed objective 1 lab and 5 red science (too easy and the following objective needs this anyway)

---------------------------------------------------------------------------------------------------
Version: 0.6.1
Date: 2023-01-22

  Balancing:
    - Objective #1: Merged stone, coal and fish objectives (too easy)
    - Objective #1: Increased stone from 30 to 50 (too easy and very needed)
    - Objective #4: Merged 30 gears and 20 electronic circuits (too easy)
    - Before objective #6: Removed 20 copper plate (too easy, needed for electronic circuits anyway)
    - Before objective #6: Removed 10 underground pipes (too opinionated / game-specific)
    - Objective #7: Increased assembling machines from 1 to 2 (bigger reward for the research)
    - Objective #15: Added 100 red science packs to the objective of 100 green science packs (bigger reward, feels more aligned with the game)

---------------------------------------------------------------------------------------------------
Version: 0.6.0
Date: 2023-01-22

  Features:
    - Added analytics and command `/ro_save_analytics` to write the analytics output to a file

---------------------------------------------------------------------------------------------------
Version: 0.5.8
Date: 2023-01-17

  Balancing:
    - Objective #30: Add 50 medium electric poles with 50 big electric poles (very needed)

---------------------------------------------------------------------------------------------------
Version: 0.5.7
Date: 2023-01-17

  Balancing:
    - Objective #35: Combine plastic and sulfur objectives (can be researched in parallel)
    - Objective #35: Increase plastic from 1 to 100 (bigger reward)
    - Objective #35: Increase sulfur from 1 to 50 (bigger reward)

---------------------------------------------------------------------------------------------------
Version: 0.5.6
Date: 2023-01-15

  Balancing:
    - Objective #24: Add 48 steel furnaces objective (next one is too hard)
    - Objective #25: Add 20 fast underground belts and 10 fast splitters to 100 fast belts objective (too easy, very needed)

---------------------------------------------------------------------------------------------------
Version: 0.5.5
Date: 2023-01-15

  Balancing:
    - Objective #20: Add a mall objective with 9 different items (very needed)

---------------------------------------------------------------------------------------------------
Version: 0.5.4
Date: 2023-01-15

  Balancing:
    - Objectives #17 and #18: Split 1000 gears and 500 green circuits (too hard)

---------------------------------------------------------------------------------------------------
Version: 0.5.3
Date: 2023-01-15

  Features:
    - Added command `/next_objective` for continuing playing after updating this mod to a new version.

---------------------------------------------------------------------------------------------------
Version: 0.5.2
Date: 2023-01-15

  Balancing:
    - Objective #16: Remove fast inserters from the objective, and mention about them in the tip (too hard)

---------------------------------------------------------------------------------------------------
Version: 0.5.1
Date: 2023-01-15

  Distribution:

control.lua
objectives = require('objectives')

function get_current_objective()
    return objectives[global.current_objective]
end

function next_objective()
    global.current_objective = (global.current_objective or 0) + 1
    if global.current_objective > #objectives then
        for _, player in pairs(game.players) do
            player.set_goal_description("") -- Hide objective pop-up
        end
    else
        for _, player in pairs(game.players) do
            update_goal_text(player, false)
        end
    end
end

commands.add_command("next_objective", nil, function(command)
    next_objective()
end)

commands.add_command("previous_objective", nil, function(command)
    if global.current_objective > 1 then
        global.current_objective = global.current_objective - 1
    end
    for _, player in pairs(game.players) do
        update_goal_text(player, false)
    end
end)

function log_analytics(line)
    global.analytics_output = (
        (global.analytics_output or "") ..
        "[0.6.6] " .. -- mod version
        "[" .. game.ticks_played .. "] " ..
        line ..
        "\n"
    )
end

commands.add_command("ro_feedback", nil, function(command)
    log_analytics("Comment: " .. command.parameter)
    game.print("Saved comment to analytics log with the current game timestamp.")
end)

commands.add_command("ro_save_analytics", nil, function(command)
    log_analytics("Saving analytics file.")
    local data = global.analytics_output
    
    local filename = "resource_objectives_analytics.txt"
    game.write_file(filename, data, false, command.player_index)
    game.print("Saved analytics to factorio/script-output/" .. filename)
end)

function is_goal_met(player)
    local objective = get_current_objective()
    local inventory = player.get_main_inventory()

    if inventory == nil then
        return false
    end

    for _, resource in pairs(objective.items) do
        local has_amount = inventory.get_item_count(resource.item)

        if has_amount < resource.quantity then
            return false
        end
    end

    log_analytics("Objective met: #" .. global.current_objective)
    return true
end

function reward(player)
    local objective = get_current_objective()
    for _, resource in pairs(objective.items) do
        game.print({
            "",
            player.name .. " earned " .. resource.quantity .. " x ",
            game.item_prototypes[resource.item].localised_name
        })
        player.insert{name=resource.item, count=resource.quantity}
    end
end

script.on_init(function()
    global.current_objective = 1
end)

function update_goal_text(player, silent)
    if global.current_objective > #objectives then
        return
    end
    local objective = get_current_objective()

    local inventory = player.get_main_inventory()

    local goal_description = {
        "", -- A special key for concatenating
    }

    local goal_append = "Have resources in your inventory:\n"

    for _, resource in pairs(objective.items) do
        local has_amount = 0
        if inventory ~= nil then
            -- Bug: the player doesn't have an inventory during the spaceship
            -- crash cut scene, so the has_amount will be 0 in the beginning of
            -- the game even if there is something. The amount will be updated
            -- once the player's inventory is changed in some way.
            has_amount = inventory.get_item_count(resource.item)
        end

        goal_append = goal_append .. "\n[item=" .. resource.item .. "] "
        table.insert(goal_description, goal_append)
        goal_append = ""

        table.insert(goal_description, game.item_prototypes[resource.item].localised_name)

        goal_append = goal_append .. ": " .. has_amount .. "/" .. resource.quantity
    end

    if objective.note then
        
        goal_append = goal_append .. "\n\n" .. objective.note
    end
    table.insert(goal_description, goal_append)

    player.set_goal_description(goal_description, silent)

    if is_goal_met(player) then
        reward(player)
        next_objective()
    end
end

function check_inventory(event)
    local player = game.players[event.player_index]
    update_goal_text(player, true)
end

function on_new_player(event)
    local player = game.players[event.player_index]
    update_goal_text(player, false)
end

script.on_event(defines.events.on_player_main_inventory_changed, check_inventory)
script.on_event(defines.events.on_player_created, on_new_player)


objectives.lua
local objectives = {
    {
        items = {
            {
                item = "stone",
                quantity = 50
            },
            {
                item = "coal",
                quantity = 30
            },
            {
                item = "raw-fish",
                quantity = 50
            }
        },
        note = "Tip: try finding some big rocks and huge rocks."
    },
    {
        items = {
            {
                item = "iron-plate",
                quantity = 50
            }
        },
        note = "Tip: when you hold a coal over a machine, you can press Z to insert one. You can also hold Z and drag the coal over multiple machines."
    },
    {
        items = {
            {
                item = "coal",
                quantity = 250
            }
        },
        note = "Tip: burner drills can feed directly into each other."
    },
    {
        items = {
            {
                item = "iron-gear-wheel",
                quantity = 50
            },
            {
                item = "electronic-circuit",
                quantity = 30
            }
        },
        note = "Tip: having intermediate materials in your inventory speeds up crafting more complex recipes."
    },
    {
        items = {
            {
                item = "stone-brick",
                quantity = 100
            }
        },
        note = "Tip: you can build a stone path with them to increase walking speed."
    },
    {
        items = {
            {
                item = "assembling-machine-1",
                quantity = 2
            }
        },
        note = "Tip: assembling machines can really speed you up, even if you hand-feed them."
    },
    {
        items = {
            {
                item = "gun-turret",
                quantity = 2
            }
        },
        note = "Tip: you might want a machine gun and armor as well."
    },
    {
        items = {
            {
                item = "firearm-magazine",
                quantity = 100
            }
        },
        note = "Tip: you can't have too many assembling machines."
    },
    {
        items = {
            {
                item = "transport-belt",
                quantity = 200
            },
            {
                item = "inserter",
                quantity = 50
            }
        }
    },
    {
        items = {
            {
                item = "iron-gear-wheel",
                quantity = 1000
            }
        },
        note = "Tip: you will need a lot more in the future, so better to make a long-term solution."
    },
    {
        items = {
            {
                item = "electronic-circuit",
                quantity = 500
            }
        }
    },
    {
        items = {
            {
                item = "gun-turret",
                quantity = 50
            },
            {
                item = "firearm-magazine",
                quantity = 1000
            }
        },
    },
    {
        items = {
            {
                item = "transport-belt",
                quantity = 500
            },
            {
                item = "underground-belt",
                quantity = 100
            },
            {
                item = "splitter",
                quantity = 50
            },
            {
                item = "inserter",
                quantity = 100
            },
            {
                item = "long-handed-inserter",
                quantity = 50
            },
            {
                item = "fast-inserter",
                quantity = 50
            },
            {
                item = "pipe",
                quantity = 100
            },
            {
                item = "assembling-machine-1",
                quantity = 50
            },
            {
                item = "electric-mining-drill",
                quantity = 50
            },
        },
        note = "Tip: make a mall."
    },
    {
        items = {
            {
                item = "automation-science-pack",
                quantity = 100
            },
            {
                item = "logistic-science-pack",
                quantity = 100
            }
        }
    },
    {
        items = {
            {
                item = "assembling-machine-2",
                quantity = 50
            },
            {
                item = "steel-furnace",
                quantity = 48
            }
        }
    },
    {
        items = {
            {
                item = "fast-transport-belt",
                quantity = 100
            },
            {
                item = "fast-underground-belt",
                quantity = 20
            },
            {
                item = "fast-splitter",
                quantity = 10
            }
        }
    },
    {
        items = {
            {
                item = "engine-unit",
                quantity = 15
            }
        }
    },
    {
        items = {
            {
                item = "car",
                quantity = 1
            }
        }
    },
    {
        items = {
            {
                item = "piercing-rounds-magazine",
                quantity = 500
            }
        }
    },
    {
        items = {
            {
                item = "military-science-pack",
                quantity = 50
            }
        }
    },
    {
        items = {
            {
                item = "big-electric-pole",
                quantity = 50
            },
            {
                item = "medium-electric-pole",
                quantity = 50
            }
        }
    },
    {
        items = {
            {
                item = "pumpjack",
                quantity = 5
            }
        }
    },
    {
        items = {
            {
                item = "flamethrower-turret",
                quantity = 5
            }
        }
    },
    {
        items = {
            {
                item = "plastic-bar",
                quantity = 100
            },
            {
                item = "sulfur",
                quantity = 50
            }
        }
    },
    {
        items = {
            {
                item = "rail",
                quantity = 1000
            },
            {
                item = "rail-signal",
                quantity = 10
            },
            {
                item = "rail-chain-signal",
                quantity = 15
            }
        }
    },
    {
        items = {
            {
                item = "advanced-circuit",
                quantity = 2000
            }
        }
    },
    {
        items = {
            {
                item = "laser-turret",
                quantity = 50
            },
            {
                item = "accumulator",
                quantity = 50
            }
        }
    },
    {
        items = {
            {
                item = "cannon-shell",
                quantity = 100
            }
        }
    },
    {
        items = {
            {
                item = "electric-engine-unit",
                quantity = 100
            }
        }
    },
    {
        items = {
            {
                item = "flying-robot-frame",
                quantity = 200
            }
        }
    },
    {
        items = {
            {
                item = "roboport",
                quantity = 4
            }
        }
    },
    {
        items = {
            {
                item = "utility-science-pack",
                quantity = 200
            }
        }
    },
    {
        items = {
            {
                item = "productivity-module-3",
                quantity = 1
            }
        }
    },
    {
        items = {
            {
                item = "space-science-pack",
                quantity = 500
            }
        }
    },
    {
        items = {
            {
                item = "artillery-shell",
                quantity = 200
            }
        }
    },
    {
        items = {
            {
                item = "atomic-bomb",
                quantity = 1
            }
        }
    },
    {
        items = {
            {
                item = "spidertron-remote",
                quantity = 1
            }
        }
    }
}

return objectives